packages/compiler/server/serverlib.ts

```diff
@@ -403,6 +410,44 @@ export function createServer(host: ServerHost): Server {
     }
   }
 
+  async function getFoldingRanges(params: FoldingRangeParams): Promise<FoldingRange[]> {
+    const file = await compilerHost.readFile(getPath(params.textDocument));
+    const ast = parse(file);
+    const ranges: FoldingRange[] = [];
+    getRangesForNode(ast);
+
+    function getRangesForNode(node: Node){
+          if ("decorators" in node && node.decorators.length > 0) {
+            const startPos = node.decorators[0].pos
+            const endPos = node.decorators[node.decorators.length-1].end
+            const nodeEndPos = node.end
+            let startLine = file.getLineAndCharacterOfPosition(startPos).line;
+            let endLine = file.getLineAndCharacterOfPosition(endPos).line;
+            if (startLine !== endLine){
+              ranges.push({startLine,endLine});
+            }
+            startLine = file.getLineAndCharacterOfPosition(skipTrivia(file.text,endPos)).line;
+            endLine = file.getLineAndCharacterOfPosition(nodeEndPos).line;
+            if (startLine !== endLine){
+              ranges.push({startLine,endLine});
+            }
+            visitChildren(node,getRangesForNode);
+            return
+          }
+          const startPos = node.pos;
+          const endPos = node.end;
+          const startLine = file.getLineAndCharacterOfPosition(startPos).line;
+          const endLine = file.getLineAndCharacterOfPosition(endPos).line;
+          if (startLine !== endLine && node.kind !== SyntaxKind.CadlScript && !(ranges.some(e =>
```

Comment:
This ranges.some call makes the algorithm O(N^2) on the number of nodes/ranges. It could slow things down dramatically in a large document. I think it would be OK to let the editor deal with dupes. If we really want to de-dupe, we should use a better data structure than re-scanning all the ranges every time we add a range.

I think it would be worthwhile to add the startChar and endChar. This will help to understand the output since there will be far fewer cases where all four coordinates are the same for more than one node. 

I'm not sure if the editor does anything interesting with the column information, but since we have it, we can provide it and see.

Reply:
I removed the range.some

Reply:
Can you also add the columns (characters) to the ranges? Check if this has any impact on the editor and if that impact looks good. And even without impact, it will help debugging to be able to tell exactly where we said something started and ended.